controller/grpc.go

```diff
@@ -0,0 +1,1014 @@
+package main
+
+import (
+	"encoding/json"
+	fmt "fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/flynn/flynn/controller/data"
+	"github.com/flynn/flynn/controller/protobuf"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/pkg/ctxhelper"
+	"github.com/flynn/flynn/pkg/postgres"
+	"github.com/flynn/flynn/pkg/random"
+	"github.com/golang/protobuf/ptypes/empty"
+	middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	log "github.com/inconshreveable/log15"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+)
+
+type grpcAPI struct {
+	*controllerAPI
+
+	db *postgres.DB
+}
+
+const ctxKeyFlynnAuthKeyID = "flynn-auth-key-id"
+
+func (g *grpcAPI) authorize(ctx context.Context) (context.Context, error) {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		if passwords, ok := md["auth-key"]; ok && len(passwords) > 0 {
+			auth, err := g.authorizer.Authorize(passwords[0])
+			if err != nil {
+				return ctx, grpc.Errorf(codes.Unauthenticated, err.Error())
+			}
+
+			if auth.ID != "" {
+				ctx = context.WithValue(ctx, ctxKeyFlynnAuthKeyID, auth.ID)
+				ctx = ctxhelper.NewContextLogger(ctx, g.logger(ctx).New("authKeyID", auth.ID))
+			}
+
+			return ctx, nil
+		}
+
+		return ctx, grpc.Errorf(codes.Unauthenticated, "no Auth-Key provided")
+	}
+
+	return ctx, grpc.Errorf(codes.Unauthenticated, "metadata missing")
+}
+
+func (g *grpcAPI) logger(ctx context.Context) log.Logger {
+	log, ok := ctxhelper.LoggerFromContext(ctx)
+	if !ok {
+		log = logger
+	}
+	return log
+}
+
+func (g *grpcAPI) logRequest(ctx context.Context, rpcMethod string) (context.Context, func(context.Context, error)) {
+	startTime := time.Now()
+
+	var clientIP string
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		var reqID []string
+		if id, ok := ctxhelper.RequestIDFromContext(ctx); ok {
+			reqID = []string{id}
+		} else {
+			reqID = md.Get("x-request-id")
+		}
+		if len(reqID) == 0 {
+			reqID = []string{random.UUID()}
+		}
+		ctx = ctxhelper.NewContextRequestID(ctx, reqID[0])
+		ctx = ctxhelper.NewContextLogger(ctx, g.logger(ctx).New("req_id", reqID[0]))
+
+		xForwardedFor := md.Get("x-forwarded-for")
+		if len(xForwardedFor) > 0 {
+			clientIPs := strings.Split(xForwardedFor[0], ",")
+			if len(clientIPs) > 0 {
+				clientIP = strings.TrimSpace(clientIPs[len(clientIPs)-1])
+			}
+		}
+	}
+	if clientIP == "" {
+		if remoteHost, ok := ctx.Value(ctxKeyRemoteHost).(string); ok {
+			clientIP = remoteHost
+		}
+	}
+
+	g.logger(ctx).Info("gRPC request started", "rpcMethod", rpcMethod, "client_ip", clientIP)
+	return ctx, func(ctx context.Context, err error) {
+		duration := time.Since(startTime)
+		if err == nil {
+			g.logger(ctx).Info("gRPC request ended", "duration", duration)
```

Comment:
I think the assumption here was that `g.logger` always print the request id so it would be fairly easy to associate it with the previous log. Though for better human readability of the logs it may be useful to include it in both.